### PR TITLE
Priority scheduling added.

### PR DESCRIPTION
**Pull request by MdAkdas**
**Fixes issue:** #4413

**Changes:**
Two files added.
priority_scheduling_non_preemptive.c
priority_scheduling_preemptive.c

### CODE REVIEW

File: code/operating_system/src/scheduling/priority_scheduling/priority_scheduling_preemptive.c [hunk 1]
```diff
@@ -0,0 +1,99 @@
+//Priority scheduling: each process is assigned a priority. Process with highest priority is to be executed first and so on.
+#include<stdio.h>
+#include<stdlib.h>
+
+struct process
+{
+    int pname, pr, at, bt;
+    int ct, wt, rt, tat, flag; //flag=0 means unfinished process flag=1 process is finished
+};
+typedef struct process process;
```
Comment (arnavb) [inline code/operating_system/src/scheduling/priority_scheduling/priority_scheduling_preemptive.c]:
```suggestion
typedef struct process
{
    int pname, pr, at, bt;
    int ct, wt, rt, tat, flag; //flag=0 means unfinished process flag=1 process is finished
} process;
```

File: code/operating_system/src/scheduling/priority_scheduling/priority_scheduling_preemptive.c [hunk 2]
```diff
@@ -0,0 +1,99 @@
+//Priority scheduling: each process is assigned a priority. Process with highest priority is to be executed first and so on.
+#include<stdio.h>
+#include<stdlib.h>
```
Comment (arnavb) [inline code/operating_system/src/scheduling/priority_scheduling/priority_scheduling_preemptive.c]:
```suggestion
#include <stdio.h>
#include <stdlib.h>
```

File: code/operating_system/src/scheduling/priority_scheduling/priority_scheduling_preemptive.c [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+process pro[20];
+float act, awt, atat, art;
+
+int next_process(void);
+
+int main(void)
+{
+    int ntemp, i;
+    printf("Enter no. of process : ");
+    scanf("%d", &n);
```
Comment (arnavb) [inline code/operating_system/src/scheduling/priority_scheduling/priority_scheduling_preemptive.c]: n should be local to this function and passed to all functions that need it. Same goes for C++ code.

File: code/operating_system/src/scheduling/priority_scheduling/priority_scheduling_preemptive.c [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+struct process
+{
+    int pname, pr, at, bt;
+    int ct, wt, rt, tat, flag; //flag=0 means unfinished process flag=1 process is finished
+};
+typedef struct process process;
+
+int t = 0, n;
+process pro[20];
+float act, awt, atat, art;
```
Comment (arnavb) [inline code/operating_system/src/scheduling/priority_scheduling/priority_scheduling_preemptive.c]: All these variables should be local and passed around as parameters, not globals.

File: code/operating_system/src/scheduling/priority_scheduling/priority_scheduling_preemptive.c [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+int next_process(void);
+
+int main(void)
+{
+    int ntemp, i;
+    printf("Enter no. of process : ");
+    scanf("%d", &n);
+    for(i = 0; i < n; i++) //input process details
```
Comment (arnavb) [inline code/operating_system/src/scheduling/priority_scheduling/priority_scheduling_preemptive.c]: Use preincrement, not post.

File: code/operating_system/src/scheduling/priority_scheduling/priority_scheduling_non_preemptive.c [hunk 6]
```diff
@@ -0,0 +1,101 @@
+//Priority scheduling: each process is assigned a priority. Process with highest priority is to be executed first and so on.
+//Non-preemptive algorithms: once a process enters the running state, it cannot be preempted until it completes its allotted time,
+
+#include<stdio.h>
+#include<stdlib.h>
```
Comment (arnavb) [inline code/operating_system/src/scheduling/priority_scheduling/priority_scheduling_non_preemptive.c]: Same review comments as your other file.

File: code/operating_system/src/scheduling/priority_scheduling/priority_scheduling_preemptive.c [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    } 
+    nprocess = i;
+
+    for(; i < n; i++)
+    {
+        if(pro[i].at <= t && pro[i].flag == 0 && pro[i].pr < pro[nprocess].pr)
+        nprocess = i;
+    }
+  return nprocess;
+}
```
Comment (arnavb) [inline code/operating_system/src/scheduling/priority_scheduling/priority_scheduling_preemptive.c]: Reformat both files with clang format and 4 spaces.

**Comment by MdAkdas** [comment]
Please review it again. @arnavb

File: code/operating_system/src/scheduling/priority_scheduling/priority_scheduling_non_preemptive.c [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        i = next_process(pro,n,t);
+        if(i > -1)
+	    {
+            printf("%d| %d |%d \t", t, pro[i].pname, t + pro[i].bt);
+            pro[i].rt = t; //setting response time
+            t += pro[i].bt;//updating time
+            pro[i].ct = t; //getting completion time
+            pro[i].flag = 1;
+            ntemp--;
+    	}
```
Comment (arnavb) [inline code/operating_system/src/scheduling/priority_scheduling/priority_scheduling_non_preemptive.c]: File still needs to be reformatted.

**Comment by MdAkdas** [comment]
@arnavb check it again. i've even squashed all the commits.

**Comment by MdAkdas** [comment]
@arnavb hi. I've got my points of the two PRs. can you please check.

**Comment by MdAkdas** [comment]
@AdiChat

**Pull request merged**